modules/geo-layers/src/imagery-layer/imagery-layer.ts

```diff
@@ -0,0 +1,208 @@
+/* global setTimeout, clearTimeout */
+
+import {
+  Layer,
+  CompositeLayer,
+  CompositeLayerProps,
+  UpdateParameters,
+  DefaultProps
+} from '@deck.gl/core';
+import {BitmapLayer} from '@deck.gl/layers';
+import type {ImageSourceMetadata, ImageType, ImageServiceType} from '@loaders.gl/wms';
+import {ImageSource, createImageSource} from '@loaders.gl/wms';
+
+export type ImageryLayerProps = CompositeLayerProps<string | ImageSource> & {
+  serviceType?: ImageServiceType | 'auto';
+  layers: string[];
+  onMetadataLoadStart: () => void;
+  onMetadataLoadComplete: (metadata: ImageSourceMetadata) => void;
+  onMetadataLoadError: (error: Error) => void;
+  onImageLoadStart: (requestId: unknown) => void;
+  onImageLoadComplete: (requestId: unknown) => void;
+  onImageLoadError: (requestId: unknown, error: Error) => void;
+};
+
+const defaultProps: ImageryLayerProps = {
+  id: 'imagery-layer',
+  data: undefined!,
+  serviceType: 'auto',
+  layers: undefined!,
+  onMetadataLoadStart: () => {},
+  onMetadataLoadComplete: () => {},
+  // eslint-disable-next-line
+  onMetadataLoadError: (error: Error) => console.error(error),
+  onImageLoadStart: () => {},
+  onImageLoadComplete: () => {},
+  // eslint-disable-next-line
+  onImageLoadError: (requestId: unknown, error: Error) => console.error(error, requestId)
+};
+
+/**
+ * The layer is used in Hex Tile layer in order to properly discard invisible elements during animation
+ */
+export class ImageryLayer extends CompositeLayer<ImageryLayerProps> {
+  static layerName = 'ImageryLayer';
+  static defaultProps: DefaultProps<ImageryLayerProps> = defaultProps;
+
+  state!: {
+    imageSource: ImageSource;
+    image: ImageType;
+    metadata: ImageSourceMetadata;
+    bounds: [number, number, number, number];
+    width: number;
+    height: number;
+  };
+
+  /** Lets deck.gl know that we want viewport change events */
+  override shouldUpdateState(): boolean {
+    return true;
+  }
+
+  override initializeState(): void {
+    // intentionally empty, initialization is done in updateState
+  }
+
+  override updateState({changeFlags, props, oldProps}: UpdateParameters<this>): void {
+    if (changeFlags.propsChanged) {
+      const dataChanged =
+        changeFlags.dataChanged ||
+        props.serviceType !== oldProps.serviceType ||
+        (changeFlags.updateTriggersChanged &&
+          (changeFlags.updateTriggersChanged.all || changeFlags.updateTriggersChanged));
+
+      // Check if data source has changed
+      if (dataChanged) {
+        this.state.imageSource = this._createImageSource(this.props);
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this._loadMetadata();
+        this.debounce(() => this.loadImage('image source changed'), 0);
+      }
+
+      // Some sublayer props may have changed
+    }
+
+    if (changeFlags.viewportChanged) {
+      this.debounce(() => this.loadImage('viewport changed'));
```

Comment:
There is a caveat when you use debounce.

`context` is shared among all layers and only correct during the call of lifecycle methods. So when a denounced function is actually executed, `context.viewport` may not be correct.

You should grab the value and pass it in:

```js
updateState(params) {
  const {viewport} = this.context;
  // ...
  this.debounce(() => this.loadImage(viewport));
}